/etc/libuser.conf Configuration Resource

djhaynes opened the pull request:
This pull request adds support for a resource to check the configuration of the /etc/libuser.conf file.

Commit:
adding etc_libuser resource

djhaynes committed:
Update resource.rb

Removed selinux resource. That will be handled in a different PR.

djhaynes committed:
Update etc_libuser.rb

Fixed errors so the tests all run correctly.

djhaynes committed:
Update etc_libuser.rb

Fixing AppVeyor errors.

adamleff commented on lib/resources/etc_libuser.rb:
```diff
@@ -0,0 +1,108 @@
+  class EtcLibUser < Inspec.resource(1)
+    name 'etc_libuser'
+    desc 'Use the etc_libuser InSpec audit resource to check the configuration of /etc/libuser.conf.
+          Due to limitations of the RSpec engine and for consistency, please use the method described here
```
I don't think this needs to be directly in the resource documentation. Perhaps you can suggest a different place for it, but this warning applies to anything that may have a `.` character in it across many resources and I would not want to litter this warning throughout all resource documentation.

adamleff commented on lib/resources/etc_libuser.rb:
```diff
@@ -0,0 +1,108 @@
+          ---end sample /etc/libuser.conf---
+
+          describe etc_libuser do
+            its(['import','login_defs']){should eq ['/etc/login.defs']}
```
Please provide spacing that is consistent with the other documentation and examples. I'd also like to request that you change this example to use the `cmp` matcher so users don't have to worry about array syntax unless necessary:

```ruby
its(['import', 'login_defs']) { should cmp '/etc/login.defs' }
```

adamleff commented on lib/resources/etc_libuser.rb:
```diff
@@ -0,0 +1,108 @@
+    class UnknownEtcLibUserConf < StandardError; end
+
+    def initialize(conf_path = nil)
+      @conf_path = conf_path || inspec.os_env('LIBUSER_CONF').content || '/etc/libuser.conf'
```
Can you provide a use case for grabbing the path via an environment variable? This is not standard in our resources and is not something we advocate for a variety of reasons.

djhaynes replied:
The reason that I added it was because the man page for /etc/libuser.conf (https://linux.die.net/man/5/libuser.conf) states:

"Files
/etc/libuser.conf The default location of the configuration file. Can be overridden by the LIBUSER_CONF environment variable, except in set-uid or set-gid programs."

adamleff commented on lib/resources/etc_libuser.rb:
```diff
@@ -0,0 +1,108 @@
+    def initialize(conf_path = nil)
+      @conf_path = conf_path || inspec.os_env('LIBUSER_CONF').content || '/etc/libuser.conf'
+      file = inspec.file(@conf_path)
+      return skip_resource "The `etc_libuser` resource is not supported on your OS. Can't find '#{@conf_path}'." unless file.file?
```
I think this should actually be a failure. If someone writes a test and the file isn't there, we cannot perform the requested assertion. We've recently changed to use exceptions for this also so it makes it easier to raise them anywhere in a resource, not just the `initialize` method.

I'd change this to:

```ruby
raise Inspec::Exceptions::FailResource, "Can't find '#{@conf_path}'" unless file.file?
```

adamleff commented on lib/resources/etc_libuser.rb:
```diff
@@ -0,0 +1,108 @@
+
+module Inspec::Resources
+  class EtcLibUser < Inspec.resource(1)
+    name 'etc_libuser'
```
On what operating systems should this resource run? We should add `supports` line(s) accordingly, such as:

```
supports os_family: 'linux'
```

... that way it won't attempt to run on Windows machines.

djhaynes replied:
Natively, it should run on Linux systems and I am happy to make this change. However, I wrote the resource such that it just checked for the config path rather than operating system because it would allow you to evaluate the configuration file on any system (since it does not require any OS-specific commands). This may not be something you want to support, but, I know it was a use case for some vendors when I worked on the OVAL project. Just let me know either way.

adamleff replied:
OK, that sounds reasonable for now, though I'm not sure it makes sense to me. :)  If we don't want to limit it to certain operating system families, I'd definitely like to see the logic to try and pull from the environment variable removed so this focuses on either the sane default, or whatever path the user supplies.

adamleff commented on lib/resources/etc_libuser.rb:
```diff
@@ -0,0 +1,108 @@
+    include ObjectTraverser
+    attr_reader :params, :raw_content
+
+    class UnknownEtcLibUserConf < StandardError; end
```
Where is this used?

adamleff commented on lib/resources/etc_libuser.rb:
```diff
@@ -0,0 +1,108 @@
+    end
+
+    def to_s
+      'Libuser Configuration'
```
Should we add the file name here in case a user test multiple libuser configs in the same profile and therefore would need to know which is which?

perhaps:

```ruby
def to_s
  "LibUser Config from #{@conf_path}"
end
```

djhaynes replied:
Yeah, I think this is a good idea. I will update the code now.

adamleff commented on lib/resources/etc_libuser.rb:
```diff
@@ -0,0 +1,108 @@
+
+    def load_raw_content(opts)
+      # if the opts isn't a hash, we assume it's a path to a file
+      unless opts.is_a?(Hash)
```
I'm unclear why this is needed. The resource is written such that only a filename parameter is expected. When would a user supply a Hash? Is this method just copy/pasta from another resource implementation? Looks like it may have been copied from the `json` resource and should probably get removed.

djhaynes replied:
Yeah, I directly copied this code from the JSON resource too. Would you like me to remove it now? Or, should I just wait until you determine the best approach for moving forward (e.g. updating SimpleConfig and subclassing INI). I am happy to do whatever you think is best. Thanks!

adamleff replied:
I believe subclassing the `ini` resource and modifying `SimpleConfig` is the right approach. So much of this resource code, including this method, will likely need to be removed. :)

adamleff commented on lib/resources/etc_libuser.rb:
```diff
@@ -0,0 +1,108 @@
+      # these are currently ResourceSkipped to maintain consistency with the resource
+      # pre-refactor (which used skip_resource). These should likely be changed to
+      # ResourceFailed during a major version bump.
+      raise Inspec::Exceptions::ResourceSkipped, "No such file: #{path}" unless file.file?
```
These should probably change to `ResourceFailed` instead of skipped... looks like this may have been copied from another resource where we skip to maintain backward compatibility, but per #2418, we plan on changing these to failures soon. Since this is a new resource, this can fail right out of the gate.

adamleff commented on lib/resources/etc_libuser.rb:
```diff
@@ -0,0 +1,108 @@
+      file.content
+    end
+
+    def load_raw_from_command(command)
```
Again, this resource isn't written to expect a user to supply a command to run to get the content of a libuser.conf file, but rather only accept a path to a file. This method should not be necessary.

adamleff commented on test/unit/resources/etc_libuser.rb:
```diff
@@ -0,0 +1,48 @@
+# encoding: utf-8
+# author: Thomas Cate
```
I don't think this is you... did Thomas Cate take part in the writing of these tests?

djhaynes replied:
No it is not. This should be fixed now.

adamleff reviewed (changes requested):
Hi, @djhaynes, and thanks for your contribution!

This looks like a great start, however it looks like a bunch of unnecessary code got copied into this resource from another resource. I've left you comments accordingly, but we definitely need to get some cleanup done before we can review again. You also have some rubocop failures which is causing the Travis failures we see.

Additionally, this project uses the [Developer Certificate of Origin](https://github.com/chef/inspec/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco) method for ensuring contributors have the legal authority to contribute. We'll need to you to either reword each of your commits to include the `Signed-off-by` line, or you can take this opportunity to squash them down to a single commit and add the `Signed-off-by` line then. But we will not be able to accept your contribution until it's signed off accordingly.

Please let us know if you have any questions. Thanks!

djhaynes committed:
Update etc_libuser.rb

fixing author name.

djhaynes committed:
Update etc_libuser.rb

Removed an unused line of code.

djhaynes committed:
Update etc_libuser.rb

Updating to_s method to indicate the path of the libuser.conf file in case multiple files are used within a profile.

Commit:
Cleaned up documentation.

Commit:
Improved error handling to use failing exceptions.

Commit:
Removing the logic to pull the libuser.conf path from an environment variable. This was done to make the resource explicit in the configuration file it takes and prevent the user from potentially getting unexpected results.
